Mathlib/NumberTheory/NumberField/Completion.lean

```diff
@@ -0,0 +1,339 @@
+  rfl
+
+/-- If the absolute value of a normed field factors through an embedding into another normed field,
+then the extended embedding `v.completion →+* A` preserves distances. -/
+theorem extensionEmbedding_dist_eq_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective)
+    (x y : v.completion) :
+    dist (extensionEmbedding_of_comp h x) (extensionEmbedding_of_comp h y) =
+      dist x y := by
+  refine (UniformSpace.Completion.induction_on₂ x y ?_ (fun x y => ?_))
+  · refine isClosed_eq ?_ continuous_dist
+    exact (continuous_iff_continuous_dist.1 (UniformSpace.Completion.continuous_extension))
+  · simp only [extensionEmbedding_of_comp_coe]
+    exact UniformSpace.Completion.dist_eq x y ▸ Isometry.dist_eq (WithAbs.isometry_of_comp h) _ _
+
+/-- If the absolute value of a normed field factors through an embedding into another normed field,
+then the extended embedding `v.completion →+* A` is an isometry. -/
+theorem isometry_extensionEmbedding_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) :
+    Isometry (extensionEmbedding_of_comp h) :=
+  Isometry.of_dist_eq <| extensionEmbedding_dist_eq_of_comp h
+
+/-- If the absolute value of a normed field factors through an embedding into another normed field,
+then the extended embedding `v.completion →+* A` is a closed embedding. -/
+theorem closedEmbedding_extensionEmbedding_of_comp
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective) :
+    ClosedEmbedding (extensionEmbedding_of_comp h) :=
+  (isometry_extensionEmbedding_of_comp h).closedEmbedding
+
+/-- If the absolute value of a normed field factors through an embedding into another normed and
+locally compact field, then the completion of the first normed field is also locally compact. -/
+theorem locallyCompactSpace [LocallyCompactSpace L]
+    (h : v = (IsAbsoluteValue.toAbsoluteValue (norm : L → ℝ)).comp f.injective)  :
+    LocallyCompactSpace (v.completion) :=
+  (closedEmbedding_extensionEmbedding_of_comp h).locallyCompactSpace
+
+end AbsoluteValue.Completion
+
+namespace NumberField.InfinitePlace
+
+open AbsoluteValue.Completion
+
+variable {K : Type*} [Field K] (v : InfinitePlace K)
+
+/-- The absolute value of an infinite place factors through its associated complex embedding. -/
+theorem abs_eq_comp :
+    v.1 = (IsAbsoluteValue.toAbsoluteValue (norm : ℂ → ℝ)).comp v.embedding.injective := by
+  rw [← v.2.choose_spec]
+  rfl
+
+/-- The absolute value of a real infinite place factors through its associated real embedding. -/
+theorem abs_of_isReal_eq_comp {v : InfinitePlace K} (hv : IsReal v) :
+    v.1 = (IsAbsoluteValue.toAbsoluteValue (norm : ℝ → ℝ)).comp
+      (v.embedding_of_isReal hv).injective := by
+  ext x
+  rw [(show v.1 x = v x by rfl), ← v.norm_embedding_of_isReal hv]
+  rfl
+
+/-- The completion of a number field at an infinite place. -/
+abbrev completion := v.1.completion
+
+namespace Completion
```

Comment:
Can you add a couple of lines as "Implementation Details" in the docstring, saying why you need to re-do the work passing from Absolute Values to Infinite Places? Have you tried to define a coercion instance from the second to the first to see if this makes it possible to write replace all the `v.1`'s simply by `v` (for an InfinitePlace `v`), so that the instances below are found automatically? A wonderful example of a file with a nice "Implementation Detail" section is Mathlib.MeasureTheory.Function.LpSpace.

Reply:
Actually, the instances are already found automatically (except for `Algebra`) so I have removed them. Defining a coercion from `InfinitePlace` to `AbsoluteValue` does not seem to make much difference, and I think the duplication is mostly now gone. Let me know if you still think an "Implementation Detail" section would be helpful (perhaps the `WithAbs` explanation could go there too if so)